![ADAPT-Logo](images/logo.png)

# ADAPTed - Adapter and poly(A) Detection And Profiling Tool

ADAPTed is a software tool that detects and profiles DNA adapters and poly(A) tails in dRNA-seq data.

Note: due to a naming conflict, the tool has been renamed from ADAPT to ADAPTed.

## Installation

To install ADAPTed, the conda/mamba environment must be installed first. To do so, you can run the following commands:

```
git clone https://github.com/wvandertoorn/ADAPTed.git
cd ADAPTed
mamba env create -f environment.yml
conda activate ADAPTed
```

Then, to install ADAPTed, run the following commands:

```
pip install .
```

## Usage

You can use ADAPTed by running the following command:

```
adapted detect --input INPUT [INPUT ...] --output OUTPUT --chemistry CHEMISTRY
```

Where `input` contains the pod5 files/directories to process, `output` specifies the location to create the output folder, and `chemistry` is 'RNA002' or 'RNA004'. Specifying the chemistry automatically selects the latest config file for that chemistry (ADAPTed currently supports `rna002_70bs` or `rna004_130bps`). Alternatively, you can specify a custom config TOML using the `--config` flag (see `adapted/config/config_files/` for examples).

For additional options, see `adapted --help` and `adapted detect --help`.

### Continue from previous run

You can continue from a previous incomplete run by using the `continue` subcommand. This is useful if the run was interrupted. When you want to continue with different performance parameters (batch size, number of processes, etc.), you can edit these variables in the `command.json` file in the output folder of the previous run prior to continuing. Take care to not overwrite any of the processing parameters, as this may cause the run to fail.

```
adapted continue /path/to/previous/run/output
```

## Input

ADAPTed supports pod5 files as input (<https://pod5-file-format.readthedocs.io/en/latest/>). These files are generated by the MinKNOW software and contain the raw signal data of the sequencing run.

## Output

- `detected_boundary_XX.csv` : contains information on detected adapter and poly(A) boundaries.
- `failed_reads_XX.csv` : contains information on reads for which no adapter was detected.
- `command.json` : contains the command line arguments used to run ADAPTed.
- `config.toml` : contains the signal processing configuration used to run ADAPTed.
- `adapted.log` : contains the log messages generated during the run.

### Result files

ADAPTed outputs 'detected_boundary_XX.csv' files, containing information on detected adapter and poly(A) boundaries.
Reads for which no adapter was detected, are described in the 'failed_reads_XX.csv' files.

These files contain the following columns:

- `read_id` : the read ID of the read.
- `read_length` : the length of the read, in number of samples in the raw signal.
- `preloaded` : the length of the preloaded signal, in number of samples in the raw signal.
- `adapter_start` : the start coordinate of the adapter signal in the raw signal.
- `adapter_end` : the end coordinate of the adapter signal, in the raw signal.
- `adapter_len` : the length of the adapter signal, in number of samples in the raw signal.
- `adapter_mean` : the mean of the adapter signal, in pico amperes.
- `adapter_std` : the standard deviation of the adapter signal, in pico amperes.
- `adapter_med` : the median of the adapter signal, in pico amperes.
- `adapter_mad` : the median absolute deviation of the adapter signal, in pico amperes.
- `polya_start` : the start coordinate of the poly(A) signal in the raw signal.
- `polya_end` : the end coordinate of the poly(A) signal, in the raw signal.
- `polya_len` : the length of the poly(A) signal, in number of samples in the raw signal.
- `polya_mean` : the mean of the poly(A) signal, in pico amperes.
- `polya_std` : the standard deviation of the poly(A) signal, in pico amperes.
- `polya_med` : the median of the poly(A) signal, in pico amperes.
- `polya_mad` : the median absolute deviation of the poly(A) signal, in pico amperes.
- `polya_truncated` : whether the poly(A) signal was truncated by the signal preload logic during detection. See [Signal preloading](#signal-preloading) for more details. When the poly(A) signal stats are of interest, consider rerunning these reads with a larger `--max_obs_trace`.
- `rna_preloaded_start` : the start coordinate of the RNA signal in the raw signal.
- `rna_preloaded_len` : the length of the (preloaded) RNA signal, in number of samples in the raw signal. This is the length of the preloaded RNA signal, which may be vastly shorter than the full read length.
- `rna_preloaded_mean` : the mean of the (preloaded) RNA signal, in pico amperes.
- `rna_preloaded_std` : the standard deviation of the (preloaded) RNA signal, in pico amperes.
- `rna_preloaded_med` : the median of the (preloaded) RNA signal, in pico amperes.
- `rna_preloaded_mad` : the median absolute deviation of the (preloaded) RNA signal, in pico amperes.
- `llr_adapter_end` : the log likelihood ratio-detected end coordinate of the adapter signal.
- `llr_rel_adapter_end` : the relative coordinate (relative to the FULL read length) of the log likelihood ratio-detected end of the adapter signal.
- `llr_adapter_end_adjust` : adjustments to the adapter end coordinate, made in llr refinement.
- `llr_polya_end_adjust` : adjustments to the poly(A) end coordinate, made in llr refinement.
- `llr_early_stop_pos` : the position of the trace calculation early stop in the raw signal, compare to `llr_adapter_end` and `preloaded` for run time performance analysis (closer to `llr_adapter_end` is better).
- `mvs_llr_polya_end_adjust_ignored` : whether the poly(A) end adjust during llr refinementwas ignored in the mean-variance-shift method.
- `mvs_llr_polya_end_to_early_stop` : if `mvs_overwrite` is `True` if can happen that the detected adapter end is downstream of the (incorrectly refined) polya-end, in that case, the polya-end is set to the `llr_early_stop_pos`. It may be wise to filter out such reads.
- `mvs_adapter_end` : the end coordinate of the adapter signal, as detected or validated using the mean-variance-shift method.
- `mvs_detect_mean_at_loc` : the local mean of the poly(A) signal.
- `mvs_detect_var_at_loc` : the local variance of the poly(A) signal.
- `mvs_detect_polya_med` : the median of the poly(A) signal.
- `mvs_detect_polya_local_range` : the local range of the poly(A) signal.
- `mvs_detect_med_shift` : the median shift of the poly(A) signal.
- `real_adapter_mean_start` : the mean of the adapter signal at the start of the adapter (window).
- `real_adapter_mean_end` : the mean of the adapter signal at the end of the adapter (window).
- `real_adapter_local_range` : the local range of the adapter signal.
- `open_pores` : the detected open pore events in the adapter signal.
- `llr_detect_log` : the log likelihood ratio detection log messages.
- `fail_reason` : the reason for the failure of the adapter detection, only present if the adapter detection failed.

## Signal preloading

ADAPTed first preloads the first N samples of the signal into memory, and then detects the adapter signal.
Sometimes, the adapter signal is longer than the preload size, this will lead to the detection failing.
Sometimes, the adapter+polyA signal is longer than the preload size, this will lead to the detection succeeding, but the polyA signal will be truncated which is of concern for downstream analysis when the length or statistics of the polyA signal are of interest.
If this is the case, we advise to run the tool once with default settings, and rerun it for truncated polyA reads with `--max_obs_trace` set to a larger value. You can use the `get_truncated.sh` script to easily obtain the truncated reads.

```
bash scripts/get_truncated.sh /directory/containing/detected/boundaries/files
adapted detect [OPTIONS] --read_id_csv /directory/containing/detected/boundaries/files/truncated_read_ids.csv --max_obs_trace VALUE
```

Depending on your poly(A) length distribution, and whether you polyadenylated your reads, choose a suitable value for `--max_obs_trace`.
A sensible value could be 1.5 or 2 times the default value, for example.

The same workflow could be applied to failed reads and may increase yield.

In all cases, it is important to note that the statistics on the rna signal (`rna_preloaded_mean`, `rna_preloaded_std`, `rna_preloaded_med`, `rna_preloaded_mad`, `rna_preloaded_len`) are based on the preloaded signal, and may not reflect the actual signal.

## Poly(A) tail length estimation

ADAPTed is capable of determining the length of the adapter and poly(A) tail in number of samples in the raw signal. This information can be combined to estimate the length of the poly(A) tail in number of bases: given the fixed length of the adapter sequence, the determined adapter length in number of samples can be used to estimate the translocation speed of the molecule. The poly(A) length in bases is then estimated by dividing the determined poly(A) length in number of samples by the translocation speed.

You can use the `resegment` workflow in our other tool ([WarpDemuX](https://github.com/KleistLab/WarpDemuX)) to obtain the median and median absolute deviation adapter event length per read.

## License

This project is licensed under the Creative Commons Attribution-NonCommercial 4.0 International License (CC BY-NC 4.0). You can view the full text of the license at the following link:
<https://creativecommons.org/licenses/by-nc/4.0/legalcode>
